File: apps/api/presentation/restapi/material_transformer.go
package restapi

import (
	"apps/api/domain"
	"encoding/json"
	"fmt"
	apiContract "libs/api-contract"
	"net/http"
	"strconv"

	"github.com/gorilla/mux"
)

func GetMaterialId(r *http.Request) (int64, error) {
	vars := mux.Vars(r)
	idParam := vars["materialId"]
	id, err := strconv.ParseInt(idParam, 10, 32)
	return id, err
}

func GetMaterialRequest(r *http.Request) (apiContract.MaterialRequest, error) {
	var materialRequest apiContract.MaterialRequest
	err := json.NewDecoder(r.Body).Decode(&materialRequest)
	return materialRequest, err
}

func ToApiMaterialSupplier(ms domain.MaterialSupplier) apiContract.MaterialSupplier {
	return apiContract.MaterialSupplier{
		Id:           ms.Id,
		SupplierId:   ms.SupplierId,
		PurchaseType: string(ms.PurchaseType),
		PurchaseUrl:  ms.PurchaseUrl,
		Supplier:     ToApiSupplier(ms.Supplier),
	}
}

func ToApiMaterial(material domain.Material) apiContract.Material {
	suppliers := []apiContract.MaterialSupplier{}
	for _, ms := range material.Suppliers {
		suppliers = append(suppliers, ToApiMaterialSupplier(ms))
	}
	return apiContract.Material{
		Id:                   material.Id,
		Name:                 material.Name,
		Price:                material.Price,
		Unit:                 material.Unit,
		WeeklyUsage:          material.WeeklyUsage,
		DeletedAt:            material.DeletedAt,
		CreatedAt:            material.CreatedAt,
		Description:          material.Description,
		PurchaseUnit:         material.PurchaseUnit,
		PurchaseUnitSize:     material.PurchaseUnitSize,
		MinimumStock:         int32(material.MinimumStock),
		NormalStock:          int32(material.NormalStock),
		Suppliers:            suppliers,
		IsStockCheckRequired: material.IsStockCheckRequired,
	}
}

func ToMaterial(materialRequest apiContract.MaterialRequest) domain.Material {
	suppliers := []domain.MaterialSupplier{}
	for _, s := range materialRequest.Suppliers {
		suppliers = append(suppliers, domain.MaterialSupplier{
			SupplierId:   s.SupplierId,
			PurchaseType: domain.PurchaseType(s.PurchaseType),
			PurchaseUrl:  s.PurchaseUrl,
		})
	}
	return domain.Material{
		Name:                 materialRequest.Name,
		Price:                materialRequest.Price,
		Unit:                 materialRequest.Unit,
		Description:          materialRequest.Description,
		PurchaseUnit:         materialRequest.PurchaseUnit,
		PurchaseUnitSize:     materialRequest.PurchaseUnitSize,
		MinimumStock:         materialRequest.MinimumStock,
		NormalStock:          materialRequest.NormalStock,
		Suppliers:            suppliers,
		IsStockCheckRequired: materialRequest.IsStockCheckRequired,
	}
}

func ValidateMaterialRequest(req apiContract.MaterialRequest) error {
	if req.PurchaseUnitSize <= 0 {
		return fmt.Errorf("purchase_unit_size must be greater than 0")
	}
	if req.MinimumStock < 0 {
		return fmt.Errorf("minimum_stock must be greater than or equal to 0")
	}
	if req.NormalStock < 0 {
		return fmt.Errorf("normal_stock must be greater than or equal to 0")
	}
	return nil
}

File: libs/ui/src/data/api/material.transformer.ts
// eslint-disable-next-line @nx/enforce-module-boundaries
import { Material as ApiMaterial } from '../../../../api-contract/src';
import { Material, MaterialForm } from '../../domain';
import { toMaterialSupplier } from './materialSupplier.transformer';

export function toMaterial(material: ApiMaterial): Material {
  return {
    id: material.id,
    createdAt: material.createdAt,
    name: material.name,
    price: material.price,
    unit: material.unit,
    description: material.description ?? '',
    weeklyUsage: material.weeklyUsage,
    purchaseUnit: material.purchaseUnit,
    purchaseUnitSize: material.purchaseUnitSize,
    minimumStock: material.minimumStock,
    normalStock: material.normalStock,
    materialSuppliers: (material.materialSuppliers ?? []).map(toMaterialSupplier),
  };
}

export function toApiMaterial(form: MaterialForm) {
  return {
    name: form.name,
    price: form.price,
    unit: form.unit,
    description: form.description,
    purchaseUnit: form.purchaseUnit,
    purchaseUnitSize: form.purchaseUnitSize,
    minimumStock: form.minimumStock,
    normalStock: form.normalStock,
    materialSuppliers: (form.materialSuppliers ?? []).map((ms) => ({
      supplierId: ms.supplierId,
      purchaseType: ms.purchaseType,
      purchaseUrl: ms.purchaseUrl,
    })),
  };
}
